### PR TITLE
fix CppcheckParserV2; allow multiple NewIssueLocations

### PR DESCRIPTION
**Pull request by ivangalkin**
This change is [Reviewable](https://reviewable.io/reviews/sonaropencommunity/sonar-cxx/1436)

### CODE REVIEW

**Comment by ivangalkin** [comment]
Dear All. Sorry for the mess with non-squashed commits. The change is ready for review. Thank you in advance!

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxReportLocation.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+  public String getFile() {
+    return file;
+  }
+
+  public String getLine() {
+    return line;
+  }
+
+  public String getMsg() {
```
Comment (guwirth) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxReportLocation.java]: Should be info

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxReportLocation.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+package org.sonar.cxx.sensors.utils;
+
+/**
+ * Each issues in SonarQube might have multiple locations; Encapsulate its
+ * properties in this structure
+ */
+public class CxxReportLocation {
+  final String file;
+  final String line;
+  final String msg;
```
Comment (guwirth) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxReportLocation.java]: should be info

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxReportLocation.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.cxx.sensors.utils;
+
+/**
+ * Each issues in SonarQube might have multiple locations; Encapsulate its
+ * properties in this structure
+ */
+public class CxxReportLocation {
```
Comment (guwirth) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxReportLocation.java]: Unit test for CxxReportLocation missing

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxReportSensor.java [hunk 4]
```diff
@@ -267,70 +269,104 @@ public static String resolveFilename(final String baseDir, @Nullable final Strin
    * @param ruleId
    * @param msg
    */
-  public void saveUniqueViolation(SensorContext sensorContext, String ruleRepoKey,
-    @Nullable String file, @Nullable String line, String ruleId, String msg) {
+  public void saveUniqueViolation(SensorContext sensorContext, String ruleRepoKey, @Nullable String file,
```
Comment (guwirth) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxReportSensor.java]: Would remove this function

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxReportSensor.java [hunk 5]
```diff
@@ -267,70 +269,104 @@ public static String resolveFilename(final String baseDir, @Nullable final Strin
...
+
+  /**
+   * Saves code violation only if unique. Compares file, line, ruleId and msg or the first given location
+   *
+   * @param sensorContext
+   * @param ruleRepoKey
+   * @param ruleId
+   * @param locations
+   */
+  public void saveUniqueViolation(SensorContext sensorContext, String ruleRepoKey, String ruleId,
```
Comment (guwirth) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxReportSensor.java]: Would call always this function from sensors.

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/cppcheck/CppcheckParserV2.java [hunk 6]
```diff
@@ -91,11 +98,19 @@ public void stream(SMHierarchicCursor rootCursor) throws XMLStreamException {
...
                   }
+
+                  CxxReportLocation location = new CxxReportLocation(file, line, (info == null) ? msg : info);
+                  locations.add(location);
                 }
 
                 if (isInputValid(id, msg)) {
-                  sensor.saveUniqueViolation(context, CxxCppCheckRuleRepository.KEY, file, line, id, msg);
+                  if (locations.isEmpty()) {
+                    sensor.saveUniqueViolation(context, CxxCppCheckRuleRepository.KEY, file, line, id, msg);
```
Comment (guwirth) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/cppcheck/CppcheckParserV2.java]: Would try to handle this always with locations

**Comment by guwirth** [comment]
@ivangalkin thanks for your proposal. My main point is that I would unify the interface and would use always the version with CxxReportLocation list only.

**Comment by ivangalkin** [comment]
Hello Günter and thank you for the reivew!

I will apply your comment in the nearest future. Unfortunately I found out, that CppcheckParserV2.java needs a bit more enhancement: For the issues with multiple locations the main message is not-equal to the message/info of the first location. E.g.
```
<error id="nullPointer" severity="warning" msg="Possible null pointer dereference: pLd" verbose="Possible null pointer dereference: pLd" cwe="476">
  <location file="file.cpp" line="598" info="Null pointer dereference"/>
  <location file="file.cpp" line="593" info="Assignment &apos;pLd=pLdFls0&apos;, assigned value is 0"/>
  <location file="file.cpp" line="569" info="Assignment &apos;pLdFls0=NULL&apos;, assigned value is 0"/>
</error>
```
According to my implementation the message becomes "Null pointer dereference", which differs from the previous logic and is quite misleading for the reader. The main message must still be "Possible null pointer dereference: pLd".

So there will be more commits.

Regards, Ivan

**Comment by ivangalkin** [comment]
I applied all review comments, but had to re-submit the change as a new pull request  #1447 (sorry, I misconfigured my fork).

**Pull request closed without merging**